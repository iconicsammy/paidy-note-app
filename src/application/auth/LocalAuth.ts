/*

  Exposes local authentication features (which depends on device)
*/

import Logger from 'application/Analytics/Logger';
import * as LocalAuthentication from 'expo-local-authentication';


export enum AuthStatus {
  CANCELLED = 'CANCELLED',
  DISABLED = 'DISABLED',
  ERROR = 'ERROR',
  SUCCESS = 'SUCCESS',
}

export interface AuthDeviceFeatures {
  fingerPrint: boolean;
  facialRecogniation: boolean;
  iris: boolean;
}

export interface AuthResult {
  success?: boolean
  error?: string

}

class LocalAuth {

  private static instance : LocalAuth;

  constructor(){
    if (!LocalAuth.instance){
      LocalAuth.instance = this;
    }
    return LocalAuth.instance
  }

  private authDeviceFeatures: AuthDeviceFeatures = {
    fingerPrint: false,
    facialRecogniation: false,
    iris: false
  }


   getSupportedAuthenticationFeatures = async (): Promise<AuthDeviceFeatures> => {
    const types = await LocalAuthentication.supportedAuthenticationTypesAsync();
    if (types && types.length) {
      this.authDeviceFeatures.facialRecogniation = types.includes(LocalAuthentication.AuthenticationType.FACIAL_RECOGNITION);
      this.authDeviceFeatures.fingerPrint =types.includes(LocalAuthentication.AuthenticationType.FINGERPRINT)
      this.authDeviceFeatures.iris = types.includes(LocalAuthentication.AuthenticationType.IRIS);
    }
    return this.authDeviceFeatures;
  };

  authenicate = async () : Promise<AuthResult> => {

      try {
        const results: AuthResult = await LocalAuthentication.authenticateAsync();
        return results;
      } catch (error) {
        Logger.logError(error);
        throw error;
      }
  }


}

export default new LocalAuth();