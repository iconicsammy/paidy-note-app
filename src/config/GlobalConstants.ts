const GlobalConstants = {
    "AUTH_FAILED": "Authentication Failed"
}

export default GlobalConstants;