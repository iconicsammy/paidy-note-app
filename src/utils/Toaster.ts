import Toast from 'react-native-root-toast';

const Toaster = {
    show(message: string){
        Toast.show(message, {
            duration: Toast.durations.LONG
        })
    }
}

export default Toaster;