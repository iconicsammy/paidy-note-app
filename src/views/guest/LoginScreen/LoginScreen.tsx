
import React from 'react';

import LocalAuth, { AuthDeviceFeatures, AuthResult } from 'application/auth/LocalAuth';
import { useEffect, useState } from 'react';
import {
    StyleSheet,
    Text,
    ImageBackground
} from "react-native";

import { Title, Button } from 'react-native-paper';
import Toaster from 'utils/Toaster'
import GlobalConstants from 'config/GlobalConstants';


function LoginScreen( {navigation }) {

    const [loggingIn, setLoggingIn] = useState(false);


    const [deviceAuthFeatures, setDeviceAuthFeatures] = useState<AuthDeviceFeatures>();

    useEffect(() => {

        const getDeviceAuthFeatures = async () => {
            const features: AuthDeviceFeatures = await LocalAuth.getSupportedAuthenticationFeatures();
            setDeviceAuthFeatures(features)
        }

        getDeviceAuthFeatures();

    }, [deviceAuthFeatures])


    const login = async () => {
        try {
            const result : AuthResult = await LocalAuth.authenicate();
            if (result.success){
                //redirect
                navigation.navigate("App")
            }else{
                //we can probe the exact error that went wrong but let's just do it here now
                // as a generic failed.
                Toaster.show(GlobalConstants.AUTH_FAILED)
            }
        } catch (error) {
            Toaster.show(GlobalConstants.AUTH_FAILED)
        }

    }


    return (

        <ImageBackground source={require("../../../../assets/stickyNote.png")}
            resizeMode="cover"
            style={styles.backgrondImage}
        >

            <Title>Note Taker</Title>

            {deviceAuthFeatures?.fingerPrint || deviceAuthFeatures?.facialRecogniation || deviceAuthFeatures?.iris ? <Button testID='btnAuth' icon="login" mode="contained" onPress={login}>
                Authenticate
            </Button> : <Text testID='txtWarning'>Your device doesn't support facial, eye or finger print auth</Text>}



        </ImageBackground>

    );
}

const styles = StyleSheet.create({
    backgrondImage: {
        flex: 1,
        marginTop: 15,
        justifyContent: 'center',
        alignItems: 'center'
    }
});

export default LoginScreen;