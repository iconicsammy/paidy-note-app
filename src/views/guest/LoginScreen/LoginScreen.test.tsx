
import React from "react";
import LoginScreen from "./LoginScreen";
import LocalAuth, { AuthDeviceFeatures, AuthResult } from 'application/auth/LocalAuth';
import { act, fireEvent, render, waitFor } from '@testing-library/react-native';
import Toaster from 'utils/Toaster'
import GlobalConstants from "config/GlobalConstants";


describe("Login Screen", () =>{
    jest.useFakeTimers();

    let spyGetSupportedAuthenticationFeatures: any;
    let spyAuthenticate: any;
    let spyToasterShow: any;

    beforeEach(()=>{
        spyGetSupportedAuthenticationFeatures = jest.spyOn(LocalAuth, "getSupportedAuthenticationFeatures");
        spyAuthenticate = jest.spyOn(LocalAuth, "authenicate");
        spyToasterShow = jest.spyOn(Toaster, "show");
    })

    afterEach(()=>{
        jest.restoreAllMocks();
    })

    it("renders correctly", async () => {
        const features: AuthDeviceFeatures = {
            fingerPrint: true,
            iris: true,
            facialRecogniation: true
        }
        spyGetSupportedAuthenticationFeatures.mockResolvedValue(features);

        const snapShotTree = await render(<LoginScreen />).toJSON();
        await act(async () => {
            expect(snapShotTree).toMatchSnapshot();
          });
      
      });

    it("should warn user device doesnt have auth features", async () =>{
        const features: AuthDeviceFeatures = {
            fingerPrint: false,
            iris: false,
            facialRecogniation: false
        }
        spyGetSupportedAuthenticationFeatures.mockResolvedValue(features);

        const { getByTestId } = render(<LoginScreen />);
        await act(async () => {
            expect(() => getByTestId('txtWarn').not.toThrow(`No instances found`));
            expect(() => getByTestId('btnAuth').toThrow(`No instances found`));
          });
      
    })


    it("should handle authentication failure", async () =>{
        const features: AuthDeviceFeatures = {
            fingerPrint: true,
            iris: false,
            facialRecogniation: false
        }
        const authResult: AuthResult = {
            success : false
        }
        spyGetSupportedAuthenticationFeatures.mockResolvedValue(features);
        spyAuthenticate.mockResolvedValue(authResult)

        const { findByTestId, getByText } = render(<LoginScreen />);
        const btnAuth = await findByTestId("btnAuth")

        await act(async () => {
            fireEvent.press(btnAuth);
          });
   
        expect(spyToasterShow).toHaveBeenCalledWith(GlobalConstants.AUTH_FAILED)
      
    })

})