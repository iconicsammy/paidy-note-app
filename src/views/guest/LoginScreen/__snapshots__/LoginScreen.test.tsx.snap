// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "marginTop": 15,
    }
  }
>
  <Image
    resizeMode="cover"
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "rgba(28, 27, 31, 1)",
        },
        Object {
          "textAlign": "left",
        },
        Array [
          Object {
            "textAlign": "left",
          },
          Object {
            "color": "rgba(28, 27, 31, 0.87)",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 20,
              "letterSpacing": 0.15,
              "lineHeight": 30,
              "marginVertical": 2,
            },
            undefined,
          ],
        ],
      ]
    }
  >
    Note Taker
  </Text>
  <Text
    testID="txtWarning"
  >
    Your device doesn't support facial, eye or finger print auth
  </Text>
</View>
`;
